File: quant1x/data/trans.cpp
#include <quant1x/data/trans.h>
#include <quant1x/level1/transaction_history.h>
#include <quant1x/cache.h>
#include <quant1x/encoding/csv.h>
#include <spdlog/spdlog.h>
#include <quant1x/factors/f10.h>
#include <filesystem>
#include <quant1x/config/cache.h>
#include <quant1x/std/filesystem.h>

namespace data {

    std::once_flag _historical_trading_data_once;
    std::mutex _historical_trading_data_mutex;
    exchange::timestamp _historical_trading_data_begin_date = exchange::timestamp(cache::trains_begin_date);

    static void lazyInitHistoricalTradingData() {
        exchange::timestamp date = exchange::timestamp(cache::trains_begin_date);
        _historical_trading_data_begin_date = date;
    }


    // 修改tick数据开始下载的日期
    static void UpdateBeginDateOfHistoricalTradingData(const std::string &date) {
        std::call_once(_historical_trading_data_once, lazyInitHistoricalTradingData);
        std::lock_guard<std::mutex> lock(_historical_trading_data_mutex);

        try {
            auto dt = exchange::timestamp(date);
            _historical_trading_data_begin_date = dt;
        } catch (...) {
            return;
        }
    }

    // 恢复默认的成交数据最早日期
    [[maybe_unused]] static void RestoreBeginDateOfHistoricalTradingData() {
        UpdateBeginDateOfHistoricalTradingData(cache::trains_begin_date);
    }

    // GetBeginDateOfHistoricalTradingData 获取系统默认的历史成交数据的最早日期
    static exchange::timestamp GetBeginDateOfHistoricalTradingData() {
        std::call_once(_historical_trading_data_once, lazyInitHistoricalTradingData);
        std::lock_guard<std::mutex> lock(_historical_trading_data_mutex);
        return _historical_trading_data_begin_date;
    }

    // 从缓存加载分笔成交数据
    static std::pair<std::vector<level1::TickTransaction>, std::string> LoadTransactionDataFromCache(const std::string &correctedCode,
                                                                                                     const exchange::timestamp &featureDate,
                                                                                                     bool ignorePreviousData) {
        std::vector<level1::TickTransaction> list;
        u32 tradeDate = featureDate.yyyymmdd();

            if (ignorePreviousData) {
            // 在默认日期之前的数据直接返回空
            auto startDate = GetBeginDateOfHistoricalTradingData();
            if (tradeDate < startDate.yyyymmdd()) {
                    spdlog::error("[data::trans] code={}, trade-date={}, start-date={}, 没有数据", correctedCode, tradeDate, startDate.toString());
                return {list, HistoricalTransactionDataFirstTime};
            }
        }

        std::string startTime = HistoricalTransactionDataFirstTime;
        std::string filename = config::get_historical_trade_filename(correctedCode, featureDate.only_date());

        if (std::filesystem::exists(filename)) {
            // 如果缓存存在
            // list = encoding::csv::csv_to_slices<level1::TickTransaction>(filename);  // 屏蔽，使用io::CSVReader解析
            io::CSVReader<6> in(filename);
            in.read_header(io::ignore_extra_column, "time", "price", "volume", "number", "amount", "buy_or_sell");
            std::string time;
            f64 price;
            i64 vol, num;
            f64 amount;
            i64 buy_or_sell;
            while (in.read_row(time, price, vol, num, amount, buy_or_sell)) {
                list.emplace_back(time, price, vol, num, amount, buy_or_sell);
            }
            if (!list.empty()) {
                size_t cacheLength = list.size();
                if (cacheLength > 0) {
                    std::string lastTime = list.back().time;
                    if (lastTime == HistoricalTransactionDataLastTime) {
                        return {list, startTime};  // 数据完整，直接返回
                    }

                    std::string firstTime;
                    size_t skipCount = 0;
                    for (size_t i = 1; i <= cacheLength; i++) {
                        std::string tm = list[cacheLength - i].time;
                        if (firstTime.empty()) {
                            firstTime = tm;
                            startTime = firstTime;
                            skipCount++;
                            continue;
                        }
                        if (tm < firstTime) {
                            startTime = firstTime;
                            break;
                        } else {
                            skipCount++;
                        }
                    }
                    // 截取 startTime之前的记录
                    list.resize(cacheLength - skipCount);
                }
            } else {
                spdlog::error("[data::trans] code={}, trade-date={}, 没有有效数据", correctedCode, tradeDate);
            }
        }

        return {list, startTime};
    }

    // 更新分笔成交数据到缓存
    static void UpdateTransactionData(const std::string &correctedCode,
                                      const exchange::timestamp &featureDate,
                                      const std::string& startTime) {
        u32 tradeDate = featureDate.yyyymmdd();
        auto today_is_last_trading_date = featureDate.is_same_date(exchange::last_trading_day());

        uint16_t offset = level1::tick_transaction_per_request_max;
        uint32_t u32Date = tradeDate;
        // 只求增量, 分笔成交数据是从后往前取数据, 缓存是从前到后顺序存取
        uint16_t start = 0;
        std::vector<level1::TickTransaction> history;
        std::vector<std::vector<level1::TickTransaction>> hs;
        auto [marketId, marketCode, pureCode] = exchange::DetectMarket(correctedCode);

        if (today_is_last_trading_date) {
            while (true) {
                level1::TransactionRequest request(correctedCode, start, offset);
                level1::TransactionResponse response(static_cast<int>(marketId), pureCode.c_str());
                auto conn = level1::get_std_conn();
                auto err = level1::process(conn->socket(), request, response);
                if (err) {
                    spdlog::error("[data::trans] code={}, tradeDate={}, error={}", correctedCode, tradeDate, std::string(err.message()));
                    break;
                }

                if (response.Count == 0 || response.List.empty()) {
                    break;
                }

                std::vector<level1::TickTransaction> tmp{};
                auto tmpList = response.List;
                std::reverse(tmpList.begin(), tmpList.end());
                for (const auto &td: tmpList) {
                    // 追加包含startTime之后的记录
                    if (td.time >= startTime) {
                        tmp.emplace_back(td);
                    }
                }
                std::reverse(tmp.begin(), tmp.end());
                hs.emplace_back(tmp);

                if (tmp.size() < offset) {
                    // 已经是最早的记录
                    // 需要排序
                    break;
                }
                start += offset;
            }
        } else {
            while (true) {
                level1::HistoryTransactionRequest request(correctedCode, u32Date, start, offset);
                level1::HistoryTransactionResponse response(static_cast<int>(marketId), pureCode.c_str());
                auto conn = level1::get_std_conn();
                auto err = level1::process(conn->socket(), request, response);
                if (err) {
                    spdlog::error("[data::trans] code={}, tradeDate={}, error={}", correctedCode, tradeDate, std::string(err.message()));
                    break;
                }

                if (response.Count == 0 || response.List.empty()) {
                    break;
                }

                std::vector<level1::TickTransaction> tmp{};
                auto tmpList = response.List;
                std::reverse(tmpList.begin(), tmpList.end());
                for (const auto &td: tmpList) {
                    // 追加包含startTime之后的记录
                    if (td.time >= startTime) {
                        tmp.emplace_back(td);
                    }
                }
                std::reverse(tmp.begin(), tmp.end());
                hs.emplace_back(tmp);

                if (tmp.size() < offset) {
                    // 已经是最早的记录
                    // 需要排序
                    break;
                }
                start += offset;
            }
        }

        // 这里需要反转一下
        std::reverse(hs.begin(), hs.end());
        for (const auto &v: hs) {
            history.insert(history.end(), v.begin(), v.end());
        }

        if (history.empty()) {
            return;  // 无新数据
        }

        // 加载现有缓存数据并合并
        auto [existingList, _] = LoadTransactionDataFromCache(correctedCode, featureDate, false);
        existingList.insert(existingList.end(), history.begin(), history.end());

        // 手动写入CSV，避免packed结构体问题
        std::string filename = config::get_historical_trade_filename(correctedCode, featureDate.only_date());
        std::string tmp = filename + ".tmp";
        auto ec = filesystem::check_filepath(tmp, true);
        ec.clear();  // 忽略错误，已处理
        {
            io::CSVWriter writer(tmp);
            // 写入header，与reader对应
            writer.write_row("time", "price", "volume", "number", "amount", "buy_or_sell");
            // 写入数据
            for (const auto& rec : existingList) {
                writer.write_row(rec.time, rec.price, rec.vol, rec.num, rec.amount, rec.buyOrSell);
            }
        }
        // 原子重命名
            try {
            std::filesystem::rename(tmp, filename);
        } catch (const std::filesystem::filesystem_error& e) {
            spdlog::error("[data::trans] 重命名失败: {} -> {}: {}", tmp, filename, e.what());
            // 尝试删除临时文件
            std::filesystem::remove(tmp);
        }
    }

    // 确保分笔成交数据已更新到缓存
    static void EnsureTransactionDataUpdated(const std::string &correctedCode,
                                             const exchange::timestamp &featureDate,
                                             bool ignorePreviousData) {
        auto [list, startTime] = LoadTransactionDataFromCache(correctedCode, featureDate, ignorePreviousData);
        bool needsUpdate = list.empty() || (list.back().time != HistoricalTransactionDataLastTime);
        if (needsUpdate) {
            UpdateTransactionData(correctedCode, featureDate, startTime);
        }
    }

    // 获取指定日期的分笔成交记录
    std::vector<level1::TickTransaction> CheckoutTransactionData(const std::string &securityCode,
                                                                 const exchange::timestamp &featureDate,
                                                                 bool ignorePreviousData) {
        std::string correctedCode = exchange::CorrectSecurityCode(securityCode);
        EnsureTransactionDataUpdated(correctedCode, featureDate, ignorePreviousData);
        auto [list, _] = LoadTransactionDataFromCache(correctedCode, featureDate, ignorePreviousData);
        return list;
    }

    // 统计指定日期的内外盘
    TurnoverDataSummary CountInflow(const std::vector<level1::TickTransaction>& list,
                                    const std::string& securityCode,
                                    const exchange::timestamp& featureDate) {
        TurnoverDataSummary summary;

        if (list.empty()) {
            return summary;
        }

        std::string correctedCode = exchange::CorrectSecurityCode(securityCode);
        double lastPrice = 0.0;

        for (const auto& v : list) {
            std::string tm = v.time;
            int64_t direction = v.buyOrSell;
            double price = v.price;

            if (lastPrice == 0) {
                lastPrice = price;
            }

            int64_t vol = v.vol;

            if (direction != level1::tick_buy && direction != level1::tick_sell) {
                if (price > lastPrice) {
                    direction = level1::tick_buy;
                } else if (price < lastPrice) {
                    direction = level1::tick_sell;
                }
            }

            // 统计内外盘数据
            if (direction == level1::tick_buy) {
                // 买入
                summary.OuterVolume += vol;
                summary.OuterAmount += static_cast<double>(vol) * price;
            } else if (direction == level1::tick_sell) {
                // 卖出
                summary.InnerVolume += vol;
                summary.InnerAmount += static_cast<double>(vol) * price;
            } else {
                // 可能存在中性盘2, 最近又发现有类型是3, 暂时还是按照中性盘来处理
                int64_t vn = vol;
                int64_t buyOffset = vn / 2;
                int64_t sellOffset = vn - buyOffset;

                // 买入
                summary.OuterVolume += buyOffset;
                summary.OuterAmount += static_cast<double>(buyOffset) * price;
                // 卖出
                summary.InnerVolume += sellOffset;
                summary.InnerAmount += static_cast<double>(sellOffset) * price;
            }

            // 计算开盘竞价数据
            if (tm >= HistoricalTransactionDataFirstTime &&
                tm < HistoricalTransactionDataStartTime) {
                summary.OpenVolume += vol;
            }

            // 计算收盘竞价数据
            if (tm > HistoricalTransactionDataFinalBiddingTime &&
                tm <= HistoricalTransactionDataLastTime) {
                summary.CloseVolume += vol;
            }

            lastPrice = price;
        }

        auto f10 = factors::get_f10(correctedCode, featureDate);
        if (f10.has_value()) {
            summary.OpenTurnZ = f10->TurnZ(f64(summary.OpenVolume));
            summary.CloseTurnZ = f10->TurnZ(f64(summary.CloseVolume));
        }

        return summary;
    }

    void DataTrans::Print(const std::string &code, const std::vector<exchange::timestamp> &dates) {
        (void)code;
        (void)dates;
    }

    void DataTrans::Update(const std::string &code, const exchange::timestamp &date) {
        std::string correctedCode = exchange::CorrectSecurityCode(code);
        EnsureTransactionDataUpdated(correctedCode, date, false);
    }
} // namespace data

File: quant1x/data/trans.rs
use crate::level1::protocol;
use crate::level1::transaction_data::{TickTransaction, TransactionRequest, TransactionResponse};
use crate::level1::{self};
use crate::Timestamp;
use std::sync::Arc;


const HISTORICAL_TRANSACTION_FIRST_TIME: &str = "09:25";
const HISTORICAL_TRANSACTION_START_TIME: &str = "09:30";
const HISTORICAL_TRANSACTION_FINAL_BIDDING_TIME: &str = "14:57";
const HISTORICAL_TRANSACTION_LAST_TIME: &str = "15:00";

#[derive(Debug)]
pub struct DataTrans;

impl crate::data::Schema for DataTrans {
    fn kind(&self) -> crate::Kind {
        crate::data::BaseTransaction
    }
    fn owner(&self) -> String {
        crate::data::DEFAULT_DATA_PROVIDER.to_string()
    }
    fn key(&self) -> String {
        "trans".to_string()
    }
    fn name(&self) -> String {
        "历史成交".to_string()
    }
    fn usage(&self) -> String {
        "历史成交".to_string()
    }
}

impl crate::data::DataAdapter for DataTrans {
    fn print(&self, _code: &str, _dates: &[Timestamp]) {}

    fn update(&self, code: &str, date: Timestamp) {
        // 遵循 C++ CheckoutTransactionData 的行为：读取缓存并分页增量拉取
        let corrected = crate::data::market::correct_security_code(code);
        let mut path = std::path::PathBuf::from(crate::config::default_cache_path());
        path.push("trans");
        // 使用按年/日期目录的组织方式：trans/YYYY/YYYY-MM-DD/<code>.csv
        let date_str = date.only_date();
        let year = if date_str.len() >= 4 {
            &date_str[..4]
        } else {
            "0000"
        };
        path.push(year);
        path.push(format!("{}", date_str));
        if let Err(e) = std::fs::create_dir_all(&path) {
            log::error!("[DataTrans] create_dir_all failed: {}", e);
            return;
        }
        path.push(format!("{}.csv", corrected));
        let filename = path.to_string_lossy().to_string();

        // 如果存在则尝试读取已有缓存(容错处理, 读取失败视为空)
        let mut list: Vec<TickTransaction> = Vec::new();
        if std::path::Path::new(&filename).exists() {
            // 尝试读取CSV；容错，读取失败视为空
            if let Ok(mut rdr) = csv::ReaderBuilder::new()
                .has_headers(true)
                .from_path(&filename)
            {
                for result in rdr.records() {
                    if let Ok(rec) = result {
                // 字段: time,price,volume,number,amount,buy_or_sell
                        let time = rec.get(0).unwrap_or("").to_string();
                        let price = rec
                            .get(1)
                            .and_then(|s| s.parse::<f64>().ok())
                            .unwrap_or(0.0);
                        let vol = rec.get(2).and_then(|s| s.parse::<i64>().ok()).unwrap_or(0);
                        let num = rec.get(3).and_then(|s| s.parse::<i64>().ok()).unwrap_or(0);
                        let amount = rec
                            .get(4)
                            .and_then(|s| s.parse::<f64>().ok())
                            .unwrap_or(0.0);
                        let buy_or_sell =
                            rec.get(5).and_then(|s| s.parse::<i64>().ok()).unwrap_or(0);
                        list.push(TickTransaction {
                            time,
                            price,
                            vol,
                            num,
                            amount,
                            buy_or_sell,
                        });
                    }
                }
            }
        }

        // 从缓存中计算起始时间（如果存在缓存）
        let mut start_time = HISTORICAL_TRANSACTION_FIRST_TIME.to_string();
        if !list.is_empty() {
            if let Some(last) = list.last() {
                if last.time == HISTORICAL_TRANSACTION_LAST_TIME {
                    return; // 已更新
                }
            }

            // 查找需要追加的最早时间（镜像C++逻辑）
            let mut first_time = String::new();
            let mut skip_count: usize = 0;
            let cache_len = list.len();
            for i in 1..=cache_len {
                let tm = &list[cache_len - i].time;
                if first_time.is_empty() {
                    first_time = tm.clone();
                    start_time = first_time.clone();
                    skip_count += 1;
                    continue;
                }
                if tm < &first_time {
                    start_time = first_time.clone();
                    break;
                } else {
                    skip_count += 1;
                }
            }
            list.truncate(cache_len - skip_count);
        }

        let today_is_last =
            date.is_same_date(&crate::exchange::last_trading_day(crate::Timestamp::now()));

        let mut start: u16 = 0;
        let mut history: Vec<TickTransaction> = Vec::new();
        let mut hs: Vec<Vec<TickTransaction>> = Vec::new();

        fn fetch_transaction_page(
            security_code: &str,
            start: u16,
            count: u16,
        ) -> Option<TransactionResponse> {
            match crate::level1::get_std_conn() {
                Ok(mut pooled) => {
                    let mut request = TransactionRequest::new(security_code, start, count);
                    let mut response = TransactionResponse::new_from_request(&request);
                    match protocol::process(pooled.stream(), &mut request, &mut response) {
                        Ok(_) => Some(response),
                        Err(e) => {
                            log::error!(
                                "level1 protocol::process error for transaction_data {}: {}",
                                security_code,
                                e
                            );
                            None
                        }
                    }
                }
                Err(e) => {
                    log::error!(
                        "failed to acquire level1 client for transaction_data {}: {}",
                        security_code,
                        e
                    );
                    None
                }
            }
        }

        if today_is_last {
            // 拉取当日实时成交分页数据
            loop {
                match fetch_transaction_page(&corrected, start, level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX) {
                    Some(mut resp) => {
                        if resp.count == 0 || resp.list.is_empty() {
                            break;
                        }
                        // C++反转每页并按start_time过滤
                        let mut tmp: Vec<TickTransaction> = Vec::new();
                        resp.list.reverse();
                        for td in resp.list.into_iter() {
                            if td.time >= start_time {
                                tmp.push(td);
                            }
                        }
                        tmp.reverse();
                        let size = tmp.len();
                        if size > 0 {
                            hs.push(tmp);
                        }
                        if (size as u16) < level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX {
                            break;
                        }
                        start = start.wrapping_add(level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX);
                    }
                    None => {
                        break;
                    }
                }
            }
        } else {
            // 获取历史成交数据页
            // 拉取历史成交分页数据
            loop {
                match crate::level1::transaction_history::fetch_history_transactions(
                    &corrected,
                    date.yyyymmdd(),
                    start,
                    level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX,
                ) {
                    Some(mut resp) => {
                        if resp.count == 0 || resp.list.is_empty() {
                            break;
                        }
                        let mut tmp: Vec<TickTransaction> = Vec::new();
                        resp.list.reverse();
                        for td in resp.list.into_iter() {
                            if td.time >= start_time {
                                tmp.push(td);
                            }
                        }
                        tmp.reverse();
                        let size = tmp.len();
                        if size > 0 {
                            hs.push(tmp);
                        }
                        if (size as u16) < level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX {
                            break;
                        }
                        start = start.wrapping_add(level1::transaction_data::TICK_TRANSACTION_PER_REQUEST_MAX);
                    }
                    None => {
                        break;
                    }
                }
            }
        }

        // 反转页面并展平
        hs.reverse();
        for v in hs.into_iter() {
            history.extend(v.into_iter());
        }

        if history.is_empty() {
            return;
        }

        list.extend(history.into_iter());

        // 写回CSV
        let tmp = format!("{}.tmp", filename);
        match std::fs::File::create(&tmp) {
            Ok(f) => {
                let mut w = csv::WriterBuilder::new().has_headers(true).from_writer(f);
                // 表头
                let header = vec!["time", "price", "volume", "number", "amount", "buy_or_sell"];
                let _ = w.write_record(&header);
                for rec in &list {
                    let _ = w.write_record(&[
                        &rec.time,
                        &rec.price.to_string(),
                        &rec.vol.to_string(),
                        &rec.num.to_string(),
                        &rec.amount.to_string(),
                        &rec.buy_or_sell.to_string(),
                    ]);
                }
                let _ = w.flush();
                if let Err(e) = std::fs::rename(&tmp, &filename) {
                    log::error!("[DataTrans] rename failed {} -> {}: {}", tmp, filename, e);
                }
            }
            Err(e) => {
                log::error!("[DataTrans] create tmp {} failed: {}", tmp, e);
            }
        }
    }
}

/// Return the full filename for a transaction cache file for `code` and `date`.
/// The date should be in "YYYYMMDD" or "YYYY-MM-DD" format.
pub fn get_trans_filepath(code: &str, date: &str) -> String {
    let corrected = crate::data::market::correct_security_code(code);
    let mut path = std::path::PathBuf::from(crate::config::default_cache_path());
    path.push("trans");
    let date_str = date.replace("-", "");
    let year = if date_str.len() >= 4 {
        &date_str[..4]
    } else {
        "0000"
    };
    path.push(year);
    path.push(&date_str);
    path.push(format!("{}.csv", corrected));
    path.to_string_lossy().to_string()
}

pub fn init() {
    let plugin = Arc::new(DataTrans) as Arc<dyn crate::data::DataAdapter>;
    crate::data::register(plugin);
}
